### ISSUE TITLE
Enhancement: [no-empty-object-type] Check for generated empty-object-type

### ISSUE THREAD

**Issue by yeonjuan** [typescript-eslint/typescript-eslint#10619]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-empty-object-type/

### Description

I suggest that the `no-empty-object-type` rule also disallows `{}` generated by type transformations.
For example, using `Omit<V,K>` on a type that is nullable will unintentionally create a `{}` type.

```ts
type NullableData = null | { name: string;  num: number };
type ITS_EMPTY_OBJECT_TYPE = Omit<NullableData, 'name'>; // '{}'
```

### Fail

```TypeScript
type Data = { name: string; num: number };
type NullableData = null | Data;

function doSomething<T extends Omit<NullableData, 'name'> // lint error
>(param: T) { }

type Unexpected = Omit<NullableData, 'name'>; // lint error
```

### Pass

```TypeScript
type Data = { name: string; num: number };

function doSomething<T extends Omit<Data, 'name'>>(param: T) { }

type Expected = Omit<Data, 'name'>;
```

### Additional Info

Perhaps it would be appropriate to handle this in no-unsafe-*?

**Comment by bradzacher** [comment]
IMO this would be a hard no because checking for this requires type information.
The rule is intended to be a fast, syntax-only check for people misusing the `{}` type -- adding type info to the rule is a breaking change that limits who can use the rule -- so generally we don't do it unless there's a REALLY good reason.

There's potential scope for this in a new rule -- but also this seems like it would be a difficult thing to check.

What does it mean to "generate" a type?
What constructs would we be checking to see if the type is `{}`?
Realistically any type operation could result in `{}` and additionally any usage of a generic type could result in `{}` -- does that mean we should check every single intersection, conditional type, and generic type reference in the codebase? That seems very slow and heavy!

**Comment by yeonjuan** [comment]
>  adding type info to the rule is a breaking change that limits who can use the rule

I see. I agree that it's not appropriate as an option of `no-empty-type`. 

Still, it would be nice if there was some other way to prevent mistakes like the one above.
The example above is actually a simplified representation of the code I found while fixing a bug.
Specifically, if the wrong type is used as a parameter to a function, it is similar to if the any type is used as a parameter, which is less type-safe.

```ts
type Data = { name: string; num: number };
type NullableData = null | Data;

function foo(param: Omit<NullableData, 'name'>) { }

foo("aaa"); // no error
foo({}); // no error
foo(1); // no error

```

> That seems very slow and heavy!

If we only check the type that is finally created when type operations are used and when Generic arguments are used, will the performance be much worse?
For nested operations, I think we can optimize by only checking the type that is finally created.

```ts
type Computed = Omit<NullableData, 'name'>; // check

function foo (arg: Computed) { // Uncheck for the references
}
```

**Comment by bradzacher** [comment]
> What if we only checked when type operations are used and when they are used as generic arguments

But how can we tell that `Omit<NullableData, 'name'>` under the hood uses an `&`?
We can't without type information!

This is the difficult problem to solve.

At its core you really want to just check every single intersection type across the codebase to look for `{}` -- which is conceptually simple enough.
However type aliases with generics mask an underlying `&` operation -- which means we can only assume that every single type reference with a generic might hide an `&` operation and thus we would need to check it.

**Comment by yeonjuan** [comment]
Okay, so as an alternative, what do you think about adding a rule like `no-unsfae-parameters` which check for `{}` is used in function & generic parameters?  This is because it implicitly accepts almost any type if “{}” is used in the parameter type, it behaves almost like `any`.

[playground](https://www.typescriptlang.org/play/?noImplicitAny=false#code/C4TwDgpgBAIghsOUC8UDeUB2cC2EBcUAzsAE4CWmA5gNxYCuOhmjARhKVAL40BQokKADl6AG1FxWoiPEQoG4qAB9YCOH14AzepgDGwcgHtMUAO6ljVAGI7dACjBxSuQgHkc5YAB4R4ydNk4ABooAHJsPFCAPigAelioAFVMCAAPSH0IABMoACI0LlzeAEp0bl5ec0sbPTtcuAbc4r4q6hr7AEYAJgBmZsqLNts7Av7+cGgAdUHrWy8oABUoNOAITCyiKCh3Tx8xCSkZNRDw3AhouITktIzVnPzC3hjUBY0BaABBeWnquY6ovjvKAAIW+M3aXgKAIqWlsBmMUE0hkMdicVGYbA4pQKvCyEF0ElI0F0xhIUDRbg83l8BwCxzCEXOMT4SJRaOalygwE2HAsnEAvBuAWR2gA)
```ts
type Data = { name: string; num: number };
type NullableData = null | Data;

function wrongFunc(param: Omit<NullableData, 'name'> // Unexpected "{}"
) { }

wrongFunc("aaa");
wrongFunc(123);
wrongFunc({});

type WrongFunc< T extends  Omit<NullableData, 'name'> // Unexpected "{}"
> = T;

type A = WrongFunc<1>;
type B = WrongFunc<{}>;


function foo(arg: number) {}
declare const arg: Omit<NullableData, 'name'> ;
foo(arg); // ts error 👍
```

**Comment by JoshuaKGoldberg** [comment]
This makes me nervous. I don't have hard data, but I suspect `{}` becomes the default inferred type parameter in more cases than we'd expect.

If someone could make a prototype version of the rule and run it on typescript-eslint & a few other large codebases, that would be helpful. I think we'll want to see real-world proof of whether this is helpful.

**Comment by yeonjuan** [comment]
> If someone could make a prototype version of the rule and run it on typescript-eslint & a few other large codebases, that would be helpful. I think we'll want to see real-world proof of whether this is helpful.

Sounds like a good idea. I'll look into it. Thanks

**Comment by yeonjuan** [comment]
Hi, @JoshuaKGoldberg 
I implemented [a roughly implemented rule](https://github.com/yeonjuan/typescript-eslint-issue-10619/blob/main/no-implicit-empty-object-type/src/rules/no-implicit-empty-object.ts) and tested it by setting it up in [typescript-eslint](https://github.com/typescript-eslint/typescript-eslint), [nest](https://github.com/nestjs/nest), [strapi](https://github.com/strapi/strapi), and [trcp](https://github.com/trpc/trpc) projects.

I couldn't find any particular errors in trcp and strapi.
An error occurred in typescript-eslint, but it's not a concern since it was caused intentionally by using the `{}` type.
(It would probably be better to ignore the operand type when an actual rule is implemented. `NonNullable<unknown>`)

https://github.com/typescript-eslint/typescript-eslint/blob/052b7f9754d81056ff81896afd3fabd132796d42/packages/eslint-plugin/src/rules/no-unsafe-assignment.ts#L368-L370


---

A meaningful error was found in the nest project. ([here](https://github.com/nestjs/nest/blob/bc8782fe475d2b99be737bc026eb930676d75e47/packages/microservices/client/client-grpc.ts#L128-L131))

```ts
    const keepaliveKeys: Record<
      keyof GrpcOptions['options']['keepalive'],
      string
    > = {
```
As shown in the changes made in [this PR](https://github.com/nestjs/nest/pull/14915), an incorrect type was used as a Record key, resulting in a `{}` type. ([simplified example on typescript playground](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAglC8UDeAoK6PoIYC4oGdgAnASwDsBzAbhQF8oAfKAVzIBMIAzciNmlUJCgBhBFABKEAMYB7ImwA8AawggZnWABooZZgFsARhCIA+fgBsIwKFLyjEARipQA9C4D87lCiA))

**Comment by JoshuaKGoldberg** [comment]
I have to say, I'm not confident that investigation points to adding type information to `no-empty-object-type` being worth the weight 😬. Obviously the `keepaliveKeys` fix is a good one, but adding type information to a rule severely limits who can use it. I think the net number of bugs caught by the rule would go down if we made it typed. Even factoring in that the % of users using typed linting is generally rising over time.

Maybe there could be a separate rule that uses types? `no-generated-empty-object-type` or some name like that? The rules could point to each other and explain the difference. @typescript-eslint/triage-team

**Comment by bradzacher** [comment]
Perhaps a separate rule, but yeah defs not in `no-empty-object-type`.
I'm kinda on the fence here just because I know that the cost of this is going to be so ridiculously high cos it'll need to analyse all types in the codebase, essentially.

**Comment by Josh-Cena** [comment]
It sounds cheaper to me than `no-unsafe-*`. You only have to analyze top-level type expressions and don't need to recurse down, for example this would be fine: `Omit<V, K> & { something: "else" }` because the intermediate  type will not be compared against any runtime value and is not unsafe.

**Comment by bradzacher** [comment]
You're not wrong that it is comparable. But I think it would be a larger scale than any one of those existing rules because you'd be checking every parameter type, every return type, every generic default and constraint, every passed generic argument, every interface property, every type alias declaration, every variable type annotation, every class property type annotation...

Then you'd also want to do some smarter analysis too - analysis beyond top-level types. Eg every tuple member type, every array type's element type.

And there's probably something to be said about cases like this:
```ts
type T = {
  prop: Omit<NullableData, 'name'>;
};
```

So you probably want to look at all object type properties as well.

It would probably end up being about as expensive as a few of the existing rules combined.

**Comment by JoshuaKGoldberg** [comment]
Coming back to this, I think it'd be useful enough to warrant a new rule. I've anecdotally experienced a few annoying types investigations where some type resolved to `{}`.

cc @typescript-eslint/triage-team WDYT?

**Comment by bradzacher** [comment]
I think there's some value -- it's just that it will be a very expensive rule as (unlike the `no-unsafe-*` rules) it needs to inspect all type reference locations and all type operations.

I'm all for this existing -- just dunno how big the _value_ is overall given the lint-time expensiveness.

**Comment by JoshuaKGoldberg** [comment]
Marking as accepting PRs - with the caveat that **if this significantly slows down typescript-eslint repository linting, we will not accept it**. In that case we would suggest that it be moved to a third-party ESLint plugin.
